frame/support/procedural/src/storage_alias.rs

```diff
@@ -0,0 +1,557 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020-2022 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: Apache-2.0
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// 	http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! Implementation of the `storage_alias` attribute macro.
+
+use frame_support_procedural_tools::generate_crate_access_2018;
+use proc_macro2::{Span, TokenStream};
+use quote::{quote, ToTokens};
+use syn::{
+	ext::IdentExt,
+	parenthesized,
+	parse::{Parse, ParseStream},
+	punctuated::Punctuated,
+	token, Attribute, Error, Ident, Result, Token, Type, TypeParam, Visibility, WhereClause,
+};
+
+/// Represents a path that only consists of [`Ident`] separated by `::`.
+struct SimplePath {
+	leading_colon: Option<Token![::]>,
+	segments: Punctuated<Ident, Token![::]>,
+}
+
+impl SimplePath {
+	/// Returns the [`Ident`] of this path.
+	///
+	/// It only returns `Some(_)` if there is exactly one element and no leading colon.
+	fn get_ident(&self) -> Option<&Ident> {
+		if self.segments.len() != 1 || self.leading_colon.is_some() {
+			None
+		} else {
+			self.segments.first()
+		}
+	}
+}
+
+impl Parse for SimplePath {
+	fn parse(input: ParseStream<'_>) -> Result<Self> {
+		Ok(Self {
+			leading_colon: if input.peek(Token![::]) { Some(input.parse()?) } else { None },
+			segments: Punctuated::parse_separated_nonempty_with(input, |p| Ident::parse_any(p))?,
+		})
+	}
+}
+
+impl ToTokens for SimplePath {
+	fn to_tokens(&self, tokens: &mut TokenStream) {
+		self.leading_colon.to_tokens(tokens);
+		self.segments.to_tokens(tokens);
+	}
+}
+
+/// Represents generics which only support [`Ident`] separated by commas as you would pass it to a
+/// type.
+struct TypeGenerics {
```

Comment:
Wouldn't this be something reusable?

Reply:
Yeah, that goes into the same direction as the discussion with @KiChjang above.